README.md
# _arcade-organizer

A Simple script to automate organizing MiSTer's \_Arcade directory based on your MRA files.

Instructions:


Download the update_arcade-organizer.sh to the Scripts directory and run.

This script looks at what MRA files you have and the information in them to organize MiSTer's \_Arcade directory. 

If the XLM tags for Year, Manufacturer, and Category are included in the MRA file, this script will create an "\_Organized" Directory in "\_Arcade" and will create the following sub-directories with soft sysmlinks to organize it:

```
_Organized
├── _1 A-E
├── _1 F-K
├── _1 L-Q
├── _1 R-T
├── _1 U-Z
├── _2 Year
├── _3 Manufacturer
└── _4 Category
```
These scripts DO NOT DUPLICATE any cores or mra files, only soft symlinks are used.

THESE SYMLINKS ONLY WORK ON MISTER! IF YOU MOUNT YOUR SD CARD OUTSIDE OF MISTER THESE SYMLINKS WILL NOT WORK.

THIS SCRIPT USES A DEFAULT LOCATION FOR `_Arcade at /media/fat/_Arcade`

AND

A DEFAULT LOCATION FOR \_Organized at `/media/fat/_Arcade/_Organized`

If your \_Arcade Directory is in a diffrent location you MUST use a `/media/fat/Scripts/update_arcade-organizer.ini` file 

If your not using the default locations for \_Arcade and \_Organized you must add their locations in `/media/fat/Scripts/update_arcade-organizer.ini` 

Q: How can I set my own custom locations for MRA and \_Organized Directories?

A: A `/media/fat/Scripts/update_arcade-organizer.ini` file may be used to set custom location for your MRA files (Scans recursivly) and \_Organized files.
Add the following line to the ini file to set a directory for MRA files: `MRADIR=/top/path/to/mra/files/_Arcade`
Add the following line to the ini file to set a directory for Organized files: `ORGDIR=/path/to/organized/files/_Organized`


Q:Will this script over write files I already have?

A: NO, This script will not clober files you already have.


Q: What If I get new MRA/Core files? 

A: You need to re-run the script to have them included in the Organized files.


**You should back up your \_Arcade directory before running this script.**

**USE AT YOUR OWN RISK - THIS COMES WITHOUT WARRANTY.**


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),

## Version 2.0 - 2021-09-17

### Added
- New directories: Resolution, Platform, Series, Move Inputs, Special Controls, Players, Num Buttons, Decades, Best-of Lists, Homebrew, Bootleg.
- Following filtering options: Regions (Main, and Other toggles), Resolution (15-24-31kHz), Rotation (0-90-180-270 ± flip), Move Inputs, Special Controls, Num Controllers, Max Num Buttons, Year (High and Low), Bootleg, Homebrew. 
- Top additional directories feature: Allows you to include the extra folders that you choose in the top level of the organized directory. Current options: Platform folders, Core folders, Year folders.
- Chronological sorting at the bottom. By enabling this, every directory will include an additional chronologically sorted index at the bottom of the list.
- Support for the MAD metadata format, which allows improving the metadata of a game without having to change the game's MRA, being thus more flexible and lightweight.
- MAD database option. The new MAD_DB ini field allows you to you use different MAD databases with different metadata (for example, different categories).
- Resources for creating your own MAD database. See: https://github.com/theypsilon/MAD_Database_MiSTer
- Toggle for activating/deactivating folders (deactivating unwanted folders will boost the speed of the Arcade Organizer): Alphabetic Dirs, Region Dir, Decades Dir, Rotation Dir, Resolution Dir, Year Dirs, Decades Dirs, Num Buttons Dir, Move Inputs Dir, Players Dir, Category Dir, Series Dir, Platform Dir, Special Controls Dir, Num Monitors Dir, Best-of Lists Dir.
- Verbose mode toggle.

### Changed
- Overall folder structure is changed to better accomodate the new folders. Main folders now are: Alphabetic Dirs, Region, Collections, Video & Inputs, Extra Software, and the customizable top additional directories.

## Version 1.0 - 2020

### Added
- Initial version


update_arcade-organizer.sh
#!/bin/bash
#set -x
#
#Simple scripts to automate organizing MiSTers _Arcade directory based on you MRA files.
#
#Instuctions:
#Download the update_arcade-organizer.sh to the Scripts directory and run.
#This script looks at what MRA files you have and the information in them to organize MiSTer's _Arcade directory.#
#If the XLM tags for Year, Manufacturer, and Category are included in the MRA file, this script will create an "_Organized" Directory in "_Arcade" and will create the following sub-directories with soft sysmlinks to organize it:
#
# _Organized
# ├── _1 A-E
# ├── _1 F-K
# ├── _1 L-Q
# ├── _1 R-T
# ├── _1 U-Z
# ├── _2 Year
# ├── _3 Manufacturer
# └── _4 Category
#
#These scripts DO NOT DUPLICATE any cores or mra files, only soft symlinks are used.
#THESE SYMLINKS ONLY WORK ON MISTER! IF YOU MOUNT YOUR SD CARD OUTSIDE OF MISTER THESE SYMLINKS WILL NOT WORK.
#
#Q: Can I set my own custom locations for MRA and _Organized Directories?
#A: A /media/fat/Scripts/update_arcade-organizer.ini file may be used to set custom location for your MRA files (Scans recursivly) and _Organized files. Add the following line to the ini file to set a directory for MRA files: MRADIR=/top/path/to/mra/files Add the following line to the ini file to set a directory for Organized files: ORGDIR=/path/to/organized/files/_Organized
#
#Q:Will this script over write files I already have?
#A: NO, This script will not clober files you already have.
#
#Q: What If I get new MRA/Core files?
#A: You need to re-run the script to have them included in the Organized files.
#
#You should back up your _Arcade directory before running this script.
#USE AT YOUR OWN RISK - THIS COMES WITHOUT WARRANTE AND MAY NEUTER EVERYTHING.
###############################################################################
# ========= OPTIONS ==================
ALLOW_INSECURE_SSL="true"
CURL_RETRY="--connect-timeout 15 --max-time 60 --retry 3 --retry-delay 5 --show-error"
# ========= CODE STARTS HERE =========

ORIGINAL_SCRIPT_PATH="${0}"
if [[ "${ORIGINAL_SCRIPT_PATH}" == "bash" ]] || [[ "${ORIGINAL_SCRIPT_PATH}" =~ ^/tmp/.*  ]]; then
    ORIGINAL_SCRIPT_PATH=$(ps | grep "^ *$PPID " | grep -o "[^ ]*$")
fi

INI_PATH="${ORIGINAL_SCRIPT_PATH%.*}.ini"
if [ -f "${INI_PATH}" ] ; then
    TMP=$(mktemp)
    dos2unix < "${INI_PATH}" 2> /dev/null | grep -v "^exit" > ${TMP} || true

    if [ $(grep -c "ALLOW_INSECURE_SSL=" "${TMP}") -gt 0 ] ; then
        ALLOW_INSECURE_SSL=$(grep "ALLOW_INSECURE_SSL=" "${TMP}" | awk -F "=" '{print$2}' | sed -e 's/^ *// ; s/ *$// ; s/^"// ; s/"$//')
    fi 2> /dev/null

    if [ $(grep -c "CURL_RETRY=" "${TMP}") -gt 0 ] ; then
        CURL_RETRY=$(grep "CURL_RETRY=" "${TMP}" | awk -F "=" '{print$2}' | sed -e 's/^ *// ; s/ *$// ; s/^"// ; s/"$//')
    fi 2> /dev/null

    rm ${TMP}
fi

SSL_SECURITY_OPTION=""

set +e
curl ${CURL_RETRY} "https://github.com" > /dev/null 2>&1
RET_CURL=$?
set -e

case ${RET_CURL} in
    0)
        ;;
    *)
        if [[ "${ALLOW_INSECURE_SSL}" == "true" ]]
        then
            SSL_SECURITY_OPTION="--insecure"
        else
            echo "CA certificates need"
            echo "to be fixed for"
            echo "using SSL certificate"
            echo "verification."
            echo "Please fix them i.e."
            echo "using security_fixes.sh"
            exit 2
        fi
        ;;
    *)
        echo "No Internet connection"
        exit 1
        ;;
esac

export SSL_SECURITY_OPTION
export CURL_RETRY

echo "STARTING: _ARCADE-ORGANIZER"
echo ""

echo "Downloading the most recent _arcade-organizer script."
echo " "
curl ${CURL_RETRY} ${SSL_SECURITY_OPTION} --location -o /tmp/_arcade-organizer.sh https://raw.githubusercontent.com/theypsilon/_arcade-organizer/master/_arcade-organizer.sh
echo

chmod +x /tmp/_arcade-organizer.sh
/tmp/_arcade-organizer.sh
rm /tmp/_arcade-organizer.sh

echo "FINISHED: _ARCADE-ORGANIZER"


rotations/regenerate_mame-rotations.sh
#!/bin/bash
#set -x
#
#Simple script to rebuild cached rotation information from MAME
#
#Instuctions:
#This is _only_ used by _arcade-organizer devs to build a 
#cached copy of mame-rotations.txt
#Which is then used at runtime for organizing the MRA files by the 
#native rotation of the screen
#
#It will clone the latest MAME repo, and then look in the MAME drivers folder
#and extract the rotational information, and cache it to a file in the local
#folder called mame-rotations.txt.  
#The intention is for this file to be committed to the organizer repo on rare occasions
#i.e. https://github.com/MAME-GETTER/_arcade-organizer
#
# by @j6wbs 2020
#
#You should back up your _Arcade directory before running this script.
#USE AT YOUR OWN RISK - THIS COMES WITHOUT WARRANTE AND MAY NEUTER EVERYTHING.
###############################################################################
echo "STARTING: REGENERATION-OF-MAME_ROTATIONS"
echo ""

echo "Cloning the MAME repo."
echo " "
git clone https://github.com/mamedev/mame.git
pushd mame
git fetch origin
LATEST_TAG="$(git describe --abbrev=0)"
echo "Version: ${LATEST_TAG}"
git checkout "${LATEST_TAG}"
popd

echo " "
echo "Extracting simple rotation information from the MAME driver files..."
grep GAME mame/src/mame/drivers/*|grep ROT |cut -d"," -f2,8|grep -v FLIP|sed -e 's/ //g'|grep ROT[0-9]*$|sort -u > mame-rotations.txt

echo " "
echo "Replacing MAME values with some overriden values (e.g. dkong appears to be ROT90 on MiSTer and not ROT270 as MAME reports)"
while read line; do sed -i "s/^`echo ${line}|cut -d, -f1`,.*/${line}/g" mame-rotations.txt; done < override-rotations.txt

echo " "
echo "Adding additional rotations"
cat additional-rotations.txt >> mame-rotations.txt
sort -o mame-rotations.txt mame-rotations.txt

echo " "
echo "Done. The new mame-rotations.txt should be ready for committing to the repo. "
echo " "
echo "Summary..."
cut -d"," -f2 mame-rotations.txt |sort|uniq -c

echo "FINISHED: _REGENERATION-OF-MAME_ROTATIONS"


.github/generate_dbs.py
#!/usr/bin/env python3
# Copyright (C) 2020, 2021 Andrew Moore "amoore2600", José Manuel Barroso Galindo "theypsilon"
# This file is part of the Arcade Organizer
#
# Arcade Organizer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Arcade Organizer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License

import sys
import subprocess
from pathlib import Path
import configparser
from inspect import currentframe, getframeinfo
import itertools
import os
import io
import hashlib
import distutils.util
import datetime
import difflib
import shutil
import time
import json
import xml.etree.cElementTree as ET

def main():

    print('START!')

    rotations = dict()
    with open('rotations/mame-rotations.txt', 'r') as rotations_file:
        for line in rotations_file:
            parts = line.split(',')
            if len(parts) == 2:
                rot = translate_mame_rotation(parts[1].strip('\n').lower())
                if rot is not None:
                    rotations[parts[0]] = rot

    subdirs_finder = SubdirsFinder('aod')
    for subdir in subdirs_finder.find_all_subdirs():
        print("\n%s:" % subdir)
        aod_finder = AodFinder(str(subdir))
        aod_reader = AodReader()

        for aod in aod_finder.find_all_aods():
            print(str(aod))
            aod_reader.read_aod(aod)
        
        data = aod_reader.data()
        for setname in rotations:
            if setname not in data:
                data[setname] = dict()

            if 'rotation' not in data[setname]:
                data[setname]['rotation'] = rotations[setname]

        json_filename = 'db/' + subdir.stem + '.json'
        zip_filename = json_filename + '.zip'
        save_data_to_compressed_json(data, json_filename, zip_filename)
        run_succesfully('git add %s' % zip_filename)

        md5_filename = zip_filename + '.md5'
        with open(md5_filename, 'w') as md5_file:
            md5_file.write(hash(zip_filename))
        run_succesfully('git add %s' % md5_filename)

    run_succesfully('git commit -m "BOT: Releasing new AOD databases." > /dev/null 2>&1 || true')

    print()
    if not run_conditional('git diff --exit-code master origin/master'):
        print("There are changes to push.")
        print()

        run_succesfully('git push origin master')
    else:
        print("Nothing to be updated.")

    print('Done.')

def translate_mame_rotation(rot):
    if rot == 'rot0':
        return 0
    elif  rot == 'rot90':
        return 90
    elif  rot == 'rot180':
        return 180
    elif  rot == 'rot270':
        return 270
    else:
        return None

def translate_aod_rotation(rot):
    if rot == 'horizontal':
        return 0
    elif  rot == 'vertical (cw)':
        return 90
    elif  rot == 'horizontal (180)':
        return 180
    elif  rot == 'vertical (ccw)':
        return 270
    else:
        return None

class AodFinder:
    def __init__(self, dir):
        self._dir = dir

    def find_all_aods(self):
        return sorted(self._scan(self._dir), key=lambda aod: aod.name.lower())

    def _scan(self, directory):
        for entry in os.scandir(directory):
            if entry.is_dir(follow_symlinks=False):
                yield from self._scan(entry.path)
            elif entry.name.lower().endswith(".aod"):
                yield Path(entry.path)

class SubdirsFinder:
    def __init__(self, dir):
        self._dir = dir

    def find_all_subdirs(self):
        return sorted(self._scan(self._dir), key=lambda subdir: subdir.name.lower())

    def _scan(self, directory):
        for entry in os.scandir(directory):
            if entry.is_dir(follow_symlinks=False):
               yield Path(entry.path)

def read_aod_fields(aod_path, tags):
    fields = { i : '' for i in tags }

    try:
        context = ET.iterparse(str(aod_path), events=("start",))
        for event, elem in context:
            elem_tag = elem.tag.lower()
            if elem_tag in tags:
                tags.remove(elem_tag)
                elem_value = elem.text
                if isinstance(elem_value, str):
                    fields[elem_tag] = elem_value
                if len(tags) == 0:
                    break
    except Exception as e:
        print("Line %s || %s (%s)" % (lineno(), e, aod_path))

    return fields

class AodReader:
    def __init__(self):
        self._data = dict()

    def read_aod(self, aod):
        fields = read_aod_fields(aod, [
            'name',
            'setname',
            'rotation',
            'flip',
            'resolution',
            'region',
            'homebrew',
            'bootleg',
            'year',
            'category',
            'manufacturer'
        ])

        data = dict()
        set_if_not_empty(data, fields, 'name')
        set_if_not_empty(data, fields, 'flip')
        set_if_not_empty(data, fields, 'resolution')
        set_if_not_empty(data, fields, 'region')
        set_if_not_empty(data, fields, 'homebrew')
        set_if_not_empty(data, fields, 'bootleg')
        set_if_not_empty(data, fields, 'year')
        set_if_not_empty(data, fields, 'category')
        set_if_not_empty(data, fields, 'manufacturer')

        if fields['rotation'] != '':
            rot = translate_aod_rotation(fields['rotation'].strip().lower())
            if rot is not None:
                data['rotation'] = rot

        self._data[fields['setname']] = data

    def data(self):
        return self._data

def set_if_not_empty(data, fields, key):
    if fields[key] != '':
        data[key] = fields[key]

def save_data_to_compressed_json(db, json_name, zip_name):

    with open(json_name, 'w') as f:
        json.dump(db, f, sort_keys=True)

    run_succesfully('touch -a -m -t 202108231405 %s' % json_name)
    run_succesfully('zip -rq -D -X -9 -A --compression-method deflate %s %s' % (zip_name, json_name))

def hash(file):
    with open(file, "rb") as f:
        file_hash = hashlib.md5()
        chunk = f.read(8192)
        while chunk:
            file_hash.update(chunk)
            chunk = f.read(8192)
        return file_hash.hexdigest()

def run_conditional(command):
    result = subprocess.run(command, shell=True, stderr=subprocess.DEVNULL, stdout=subprocess.PIPE)

    stdout = result.stdout.decode()
    if stdout.strip():
        print(stdout)
        
    return result.returncode == 0

def run_succesfully(command):
    result = subprocess.run(command, shell=True, stderr=subprocess.PIPE, stdout=subprocess.PIPE)

    stdout = result.stdout.decode()
    stderr = result.stderr.decode()
    if stdout.strip():
        print(stdout)
    
    if stderr.strip():
        print(stderr)

    if result.returncode != 0:
        raise Exception("subprocess.run Return Code was '%d'" % result.returncode)

if __name__ == '__main__':
    main()